[red-knot] More precise inference for chained boolean expressions (#15089)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/narrow/truthiness.md

# TODO: It should be A. We should improve UnionBuilder or IntersectionBuilder. (issue #15023)
reveal_type(y)  # revealed: A & ~AlwaysTruthy | A & ~AlwaysFalsy
```

> ## Filtering in chained boolean expressions

Reviewer: The term we use for this is "narrowing", not "filtering"
```suggestion
## Narrowing in chained boolean expressions
```

```py
from typing import Literal

class A: ...

def _(x: Literal[0, 1]):
    reveal_type(x or A())  # revealed: Literal[1] | A
    reveal_type(x and A())  # revealed: Literal[0] | A

> # Filtering should not happen in cases where it makes the result type more complex
> def _(x: str):
>     reveal_type(x or A())  # revealed: str | A
>     reveal_type(x and A())  # revealed: str | A

Reviewer: This is not part of the requirements in the linked task. For the scope of this task, we should simply intersect with `~AlwaysTruthy` or `~AlwaysFalsy` and let the resulting types shake out.

If we feel we need to simplify display of some types to avoid user confusion, we may do that in future, but we want to evaluate that holistically in the context of real-world code, not pre-emptively. And we will implement that in a general way (either in type display or in intersection building), not special-cased to a single narrowing scenario.
```suggestion
def _(x: str):
    reveal_type(x or A())  # revealed: str & ~AlwaysFalsy | A
    reveal_type(x and A())  # revealed: str & ~AlwaysTruthy | A
```

def _(x: bool | str):
    reveal_type(x or A())  # revealed: Literal[True] | str | A
    reveal_type(x and A())  # revealed: Literal[False] | str | A

class Falsy:
    def __bool__(self) -> Literal[False]: ...

class Truthy:
    def __bool__(self) -> Literal[True]: ...

def _(x: Falsy | Truthy):
    reveal_type(x or A())  # revealed: Truthy | A
    reveal_type(x and A())  # revealed: Falsy | A

class MetaFalsy(type):
    def __bool__(self) -> Literal[False]: ...

class MetaTruthy(type):
    def __bool__(self) -> Literal[False]: ...

class FalsyClass(metaclass=MetaFalsy): ...
class TruthyClass(metaclass=MetaTruthy): ...

def _(x: type[FalsyClass] | type[TruthyClass]):
    # TODO: Should be `type[TruthyClass] | A`
    reveal_type(x or A())  # revealed: type[FalsyClass] | type[TruthyClass] | A
    # TODO: Should be `type[FalsyClass] | A`

>     reveal_type(x and A())  # revealed: type[FalsyClass] | type[TruthyClass] | A

Reviewer: FWIW, in case you're interested, I don't think this should be a difficult TODO to address (and I think it will also fix some other TODOs in the tests.) It just requires handling `ClassLiteral` differently in `Type::bool`. (The TODO comment there says we should look up "`__bool__` and `__len__`" methods on the metaclass, but this is wrong, we've since realized we can't rely on `__len__` for anything, as explained below under the Instance case.)

In fact I think the simplest implementation might be to treat the `ClassLiteral` type as instead an `Instance` type of its metaclass, and just recursively defer to the `Instance` case.

Probably best to do this in a separate PR though.
